File: src/transform.py
"""
Converts raw FPL API dicts into typed model instances.
All data cleaning and normalisation lives here, keeping models.py and
database.py free of business logic.
"""
from __future__ import annotations

import logging
from typing import Any

from src.models import (
    Fixture,
    Gameweek,
    LiveGameweekStats,
    Player,
    PlayerHistory,
    PlayerHistoryPast,
    Team,
)

logger = logging.getLogger(__name__)


def transform_bootstrap(
    data: dict[str, Any],
) -> tuple[list[Team], list[Gameweek], list[Player]]:
    """
    Parse the bootstrap-static response into Teams, Gameweeks, and Players.

    Args:
        data: Raw dict from FPLAPI.get_bootstrap_static()

    Returns:
        (teams, gameweeks, players)
    """
    teams: list[Team] = []
    for raw in data.get("teams", []):
        try:
            teams.append(Team.from_dict(raw))
        except Exception as exc:
            logger.warning("Skipping malformed team %s: %s", raw.get("id"), exc)

    gameweeks: list[Gameweek] = []
    for raw in data.get("events", []):
        try:
            gameweeks.append(Gameweek.from_dict(raw))
        except Exception as exc:
            logger.warning("Skipping malformed gameweek %s: %s", raw.get("id"), exc)

    players: list[Player] = []
    for raw in data.get("elements", []):
        try:
            players.append(Player.from_dict(raw))
        except Exception as exc:
            logger.warning("Skipping malformed player %s: %s", raw.get("id"), exc)

    logger.debug(
        "Transformed bootstrap: %d teams, %d gameweeks, %d players",
        len(teams), len(gameweeks), len(players),
    )
    return teams, gameweeks, players


def transform_element_summary(
    player_fpl_id: int,
    data: dict[str, Any],
) -> tuple[list[PlayerHistory], list[PlayerHistoryPast]]:
    """
    Parse an element-summary response for a single player.

    Args:
        player_fpl_id: The FPL ID of the player this summary belongs to.
        data: Raw dict from FPLAPI.get_element_summary(player_id)

    Returns:
        (history, history_past)
    """
    history: list[PlayerHistory] = []
    for raw in data.get("history", []):
        try:
            history.append(PlayerHistory.from_dict(player_fpl_id, raw))
        except Exception as exc:
            logger.warning(
                "Skipping malformed history row for player %d (GW %s): %s",
                player_fpl_id, raw.get("round"), exc,
            )

    history_past: list[PlayerHistoryPast] = []
    for raw in data.get("history_past", []):
        try:
            history_past.append(PlayerHistoryPast.from_dict(player_fpl_id, raw))
        except Exception as exc:
            logger.warning(
                "Skipping malformed history_past row for player %d (%s): %s",
                player_fpl_id, raw.get("season_name"), exc,
            )

    return history, history_past


def transform_fixtures(data: list[dict[str, Any]]) -> list[Fixture]:
    """
    Parse the fixtures list response.

    Args:
        data: Raw list from FPLAPI.get_fixtures()
    """
    fixtures: list[Fixture] = []
    for raw in data:
        try:
            fixtures.append(Fixture.from_dict(raw))
        except Exception as exc:
            logger.warning("Skipping malformed fixture %s: %s", raw.get("id"), exc)
    logger.debug("Transformed %d fixtures", len(fixtures))
    return fixtures


def transform_event_live(
    gameweek_fpl_id: int,
    data: dict[str, Any],
) -> list[LiveGameweekStats]:
    """
    Parse the event/{gw}/live response.

    Args:
        gameweek_fpl_id: The gameweek number (used to populate the foreign key).
        data: Raw dict from FPLAPI.get_event_live(gameweek)
    """
    rows: list[LiveGameweekStats] = []
    for element in data.get("elements", []):
        try:
            player_fpl_id = int(element["id"])
            rows.append(
                LiveGameweekStats.from_dict(player_fpl_id, gameweek_fpl_id, element)
            )
        except Exception as exc:
            logger.warning(
                "Skipping malformed live stats for element %s: %s",
                element.get("id"), exc,
            )
    logger.debug("Transformed %d live stats rows for GW%d", len(rows), gameweek_fpl_id)
    return rows

File: src/transform.ts
import { getLogger } from "./logger.ts";
import {
  Fixture,
  Gameweek,
  LiveGameweekStats,
  Player,
  PlayerHistory,
  PlayerHistoryPast,
  Team,
} from "./models.ts";

const logger = getLogger("src.transform");

type RawRecord = Record<string, unknown>;

function isRecord(value: unknown): value is RawRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/** Build the bootstrap model rows consumed by sync and persistence without duplicating normalization. */
export function transform_bootstrap(
  data: unknown,
): [Team[], Gameweek[], Player[]] {
  const payload = isRecord(data) ? data : {};

  const teams: Team[] = [];
  const rawTeams = Array.isArray(payload.teams) ? payload.teams : [];
  for (const raw of rawTeams) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed team %s: expected object payload",
        undefined,
      );
      continue;
    }

    try {
      teams.push(Team.fromDict(raw));
    } catch (error) {
      logger.warn("Skipping malformed team %s: %s", raw.id, error);
    }
  }

  const gameweeks: Gameweek[] = [];
  const rawGameweeks = Array.isArray(payload.events) ? payload.events : [];
  for (const raw of rawGameweeks) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed gameweek %s: expected object payload",
        undefined,
      );
      continue;
    }

    try {
      gameweeks.push(Gameweek.fromDict(raw));
    } catch (error) {
      logger.warn("Skipping malformed gameweek %s: %s", raw.id, error);
    }
  }

  const players: Player[] = [];
  const rawPlayers = Array.isArray(payload.elements) ? payload.elements : [];
  for (const raw of rawPlayers) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed player %s: expected object payload",
        undefined,
      );
      continue;
    }

    try {
      players.push(Player.fromDict(raw));
    } catch (error) {
      logger.warn("Skipping malformed player %s: %s", raw.id, error);
    }
  }

  logger.debug(
    "Transformed bootstrap: %d teams, %d gameweeks, %d players",
    teams.length,
    gameweeks.length,
    players.length,
  );
  return [teams, gameweeks, players];
}

/** Build the per-player history rows for sync/database consumers while skipping bad summary entries. */
export function transform_element_summary(
  player_fpl_id: number,
  data: unknown,
): [PlayerHistory[], PlayerHistoryPast[]] {
  const payload = isRecord(data) ? data : {};

  const history: PlayerHistory[] = [];
  const rawHistory = Array.isArray(payload.history) ? payload.history : [];
  for (const raw of rawHistory) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed history row for player %d (GW %s): %s",
        player_fpl_id,
        undefined,
        "expected object payload",
      );
      continue;
    }

    try {
      history.push(PlayerHistory.fromDict(player_fpl_id, raw));
    } catch (error) {
      logger.warn(
        "Skipping malformed history row for player %d (GW %s): %s",
        player_fpl_id,
        raw.round,
        error,
      );
    }
  }

  const history_past: PlayerHistoryPast[] = [];
  const rawHistoryPast = Array.isArray(payload.history_past)
    ? payload.history_past
    : [];
  for (const raw of rawHistoryPast) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed history_past row for player %d (%s): %s",
        player_fpl_id,
        undefined,
        "expected object payload",
      );
      continue;
    }

    try {
      history_past.push(PlayerHistoryPast.fromDict(player_fpl_id, raw));
    } catch (error) {
      logger.warn(
        "Skipping malformed history_past row for player %d (%s): %s",
        player_fpl_id,
        raw.season_name,
        error,
      );
    }
  }

  return [history, history_past];
}

/** Build stable fixture rows from raw fixture collections so downstream ports can persist them directly. */
export function transform_fixtures(data: unknown): Fixture[] {
  const rawFixtures = Array.isArray(data) ? data : [];
  const fixtures: Fixture[] = [];

  for (const raw of rawFixtures) {
    if (!isRecord(raw)) {
      logger.warn(
        "Skipping malformed fixture %s: %s",
        undefined,
        "expected object payload",
      );
      continue;
    }

    try {
      fixtures.push(Fixture.fromDict(raw));
    } catch (error) {
      logger.warn("Skipping malformed fixture %s: %s", raw.id, error);
    }
  }

  logger.debug("Transformed %d fixtures", fixtures.length);
  return fixtures;
}

/** Build live-stat rows for one gameweek so later sync/database ports can upsert them unchanged. */
export function transform_event_live(
  gameweek_fpl_id: number,
  data: unknown,
): LiveGameweekStats[] {
  const payload = isRecord(data) ? data : {};
  const rawElements = Array.isArray(payload.elements) ? payload.elements : [];
  const rows: LiveGameweekStats[] = [];

  for (const element of rawElements) {
    if (!isRecord(element)) {
      logger.warn(
        "Skipping malformed live stats for element %s: %s",
        undefined,
        "expected object payload",
      );
      continue;
    }

    try {
      const player_fpl_id = Number.parseInt(String(element.id), 10);
      if (!Number.isSafeInteger(player_fpl_id)) {
        throw new TypeError("Expected integer for element.id");
      }

      rows.push(
        LiveGameweekStats.fromDict(player_fpl_id, gameweek_fpl_id, element),
      );
    } catch (error) {
      logger.warn(
        "Skipping malformed live stats for element %s: %s",
        element.id,
        error,
      );
    }
  }

  logger.debug(
    "Transformed %d live stats rows for GW%d",
    rows.length,
    gameweek_fpl_id,
  );
  return rows;
}
